design/ceph-mon-pv.md

```diff
@@ -0,0 +1,97 @@
+`StorageClassName` and `Resources`. Rook follows the standard convention of
+using the default storage class when one is not specified in a volume claim
+template. If the storage resource requirements are not specified in the claim
+template, then Rook will use a default value. This is possible because unlike
+the storage requirements of OSDs (xf: StorageClassDeviceSets), reasonable
+defaults (e.g. 5-10 GB) exist for monitor daemon storage needs.
+
+*Logs and crash data*. The current implementation continues the use of a
+HostPath volume based on `dataDirHostPath` for storing daemon log and crash
+data. This is a temporary exception that will be resolved as we converge on an
+approach that works for all Ceph daemon types.
+
+Finally, the entire volume claim template may be left unspecified in the CRD
+in which case the existing HostPath mechanism is used for all monitor storage.
+
+## Upgrades and CRD changes
+
+When a new monitor is created it uses the _current_ storage specification found
+in the CRD. Once a monitor has been created, it's backing storage is not
+changed. This makes upgrades particularly simple because existing monitors
+continue to use the same storage.
+
+Once a volume claim template is defined in the CRD new monitors will be created
+with PVC storage. In order to remove old monitors based on HostPath storage
+first define a volume claim template in the CRD and then fail over each monitor.
+
+## Clean up
+
+Like `StatefulSet`s removal of an monitor deployment does not automatically
+remove the underlying PVC. This is a safety mechanism so that the data is not
+automatically destroyed. The PVCs can be removed manually once the cluster is
+healthy.
+
+## Requirements and configuration
+
+Rook currently makes explicit scheduling decisions for monitors by using node
+selectors to force monitor node affinity. This means that the volume binding
+should not occur until the pod is scheduled onto a specific node in the cluster.
+This should be done by using the `WaitForFirstConsumer` binding policy on the
+storage class used to provide PVs to monitors:
+
+```
+kind: StorageClass
+volumeBindingMode: WaitForFirstConsumer
+```
+
+When using existing HostPath storage or non-local PVs that can migrate (e.g.
+network volumes like RBD or EBS) existing monitor scheduling will continue to
+work as expected. However, because monitors are scheduled without considering
+the set of available PVs, when using statically provisioned local volumes Rook
+expects volumes to be available. Therefore, when using locally provisioned
+volumes take care to ensure that each node has storage provisioned.
+
+Note that these limitations are currently imposed because of the explicit
+scheduling implementation in Rook. These restrictions will be removed or
+significantly relaxed once monitor scheduling is moved under control of Kubernetes
+itself (on going work).
+
+## Migration
+
+Monitor migration is currently not implemented. When nodes are decommissioned,
```

Comment:
Is this defeating the purpose of having the PVCs if we never get to reuse the PVC to reschedule pods? Or am I reading this wrong?

This does make sense for host networking.

Reply:
It makes sense to use PVC storage even when migration is not possible simply because local storage may not be available. However, this patch disables migration for all cases out of simplicity. Enabling migration for a strict subset of the deployment cases is a new PR being worked on right now.